### PR TITLE
logging exceptions

### PR DESCRIPTION
**Pull request by ymatsuda**
@guozhangwang

### CODE REVIEW

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java [hunk 1]
```diff
@@ -148,11 +148,30 @@ public void close() {
...
         log.info("Shutting down a stream thread");
-        commitAll(time.milliseconds());
 
-        producer.close();
-        consumer.close();
-        removePartitions();
+        // Exceptions should not prevent this call from going through all shutdown steps.
+
+        try {
+            commitAll(time.milliseconds());
```
Comment (guozhangwang) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java]: Do we still need to re-throw in commitAll at all if they are not handled here?

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java [hunk 2]
```diff
@@ -148,11 +148,30 @@ public void close() {
...
+        }
+        try {
+            consumer.close();
+        } catch (Throwable e) {
+            log.error("failed to close consumer: ", e);
+        }
+        try {
+            removePartitions();
+        } catch (Throwable e) {
+            // already logged in removePartition()
```
Comment (guozhangwang) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java]: Ditto above.

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java [hunk 3]
```diff
@@ -285,7 +313,8 @@ private void removePartitions() {
             try {
                 task.close();
             } catch (Exception e) {
-                throw new KafkaException(e);
+                log.error("failed to close a task: ", e);
```
Comment (guozhangwang) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java]: include task id in error log.

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java [hunk 4]
```diff
@@ -268,8 +292,12 @@ private void addPartitions(Collection<TopicPartition> assignment) {
                         partitionsForTask.add(part);
 
                 // create the task
-                task = new StreamTask(id, consumer, producer, partitionsForTask, builder.build(), config);
-
+                try {
+                    task = new StreamTask(id, consumer, producer, partitionsForTask, builder.build(), config);
+                } catch (Exception e) {
+                    log.error("failed to create a task: ", e);
```
Comment (guozhangwang) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java]: Include task id in log message?

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java [hunk 5]
```diff
@@ -218,7 +237,12 @@ private void maybeCommit() {
      */
     private void commitAll(long now) {
         for (StreamTask task : tasks.values()) {
-            task.commit();
+            try {
+                task.commit();
+            } catch (Exception e) {
+                log.error("failed to commit: ", e);
```
Comment (guozhangwang) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java]: Ditto below

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java [hunk 6]
```diff
@@ -148,11 +148,30 @@ public void close() {
...
+
+        try {
+            commitAll(time.milliseconds());
+        } catch (Throwable e) {
+            // already logged in commitAll()
+        }
+        try {
+            producer.close();
+        } catch (Throwable e) {
+            log.error("failed to close producer: ", e);
```
Comment (guozhangwang) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java]: Could we expose the thread id and include it in the logging message?

**Comment by guozhangwang** [comment]
Could we make the logging message first letter capitalized?

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java (hunk 1 above)
Comment (ymatsuda) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java]: commitAll() is also called from onPartitionsRevoked(). I am not sure how KafkaConsumer handles a runtime exception from ConsumerRebalanceCallback.

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java (hunk 6 above)
Comment (ymatsuda) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java]: sure

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java (hunk 2 above)
Comment (ymatsuda) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java]: sure

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java (hunk 5 above)
Comment (ymatsuda) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java]: sure

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java (hunk 4 above)
Comment (ymatsuda) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java]: sure

File: stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java (hunk 3 above)
Comment (ymatsuda) [inline stream/src/main/java/org/apache/kafka/streaming/processor/internals/StreamThread.java]: sure

**Pull request merged**